pkg/controller/scheduler.go

```diff
@@ -247,7 +251,72 @@ func (c *Controller) advanceCanary(name string, namespace string, skipLivenessCh
 		}
 	}
 
-	// increase canary traffic percentage
+	// canary fix routing: A/B testing
+	if len(cd.Spec.CanaryAnalysis.Match) > 0 {
+		// route traffic to canary and increment iterations
+		if cd.Spec.CanaryAnalysis.Iterations > cd.Status.Iterations {
+			if err := meshRouter.SetRoutes(cd, 0, 100); err != nil {
```

Comment:
Question here. Why do we shift 100% of traffic into the canary? I thought only a certain amount of users with match headers/cookies will get the canary build, while the rest will still get the primary build. Wouldn't doing this make all users to use the canary traffic?

Reply:
In this case the 100% traffic is for the match conditions. Here is the generated virtual service:

```yaml
apiVersion: networking.istio.io/v1alpha3
kind: VirtualService
metadata:
  creationTimestamp: "2019-03-08T00:14:48Z"
  generation: 1
  name: podinfo
  namespace: test
  ownerReferences:
  - apiVersion: flagger.app/v1alpha3
    blockOwnerDeletion: true
    controller: true
    kind: Canary
    name: podinfo
    uid: 2e19ad54-4137-11e9-8e1d-42010a9c0fd1
  resourceVersion: "7525559"
  selfLink: /apis/networking.istio.io/v1alpha3/namespaces/test/virtualservices/podinfo
  uid: 2f5d0a5a-4137-11e9-8e1d-42010a9c0fd1
spec:
  gateways:
  - public-gateway.istio-system.svc.cluster.local
  - mesh
  hosts:
  - app.istio.weavedx.com
  - podinfo
  http:
  - appendHeaders:
      x-envoy-max-retries: "10"
      x-envoy-retry-on: gateway-error,connect-failure,refused-stream
      x-envoy-upstream-rq-timeout-ms: "15000"
    match:
    - uri:
        prefix: /
    - headers:
        user-agent:
          regex: "^(?!.*Chrome)(?=.*\bSafari\b).*$"
    - headers:
        cookie:
          regex: ^(.*?;)?(user=test)(;.*)?$
    rewrite:
      uri: /
    route:
    - destination:
        host: podinfo-primary
        port:
          number: 9898
      weight: 100
    - destination:
        host: podinfo-canary
        port:
          number: 9898
      weight: 0
  - appendHeaders:
      x-envoy-max-retries: "10"
      x-envoy-retry-on: gateway-error,connect-failure,refused-stream
      x-envoy-upstream-rq-timeout-ms: "15000"
    match:
    - uri:
        prefix: /
    rewrite:
      uri: /
    route:
    - destination:
        host: podinfo-primary
        port:
          number: 9898
      weight: 100
```

Reply:
ahh understood this now. Thanks!